HFE.md

```diff
@@ -0,0 +1,847 @@
+* Performance impact - although the trie hight is bounded, when encapsulating large number of items the trie is expected to hold many internal nodes. this will require up to 16 dereferences between different nodes allocated to support the trie structure which can lead to impact on performance on the regular path (when performing hash inserts, deleted and TTL updates) for items holding TTL.
+
+Although we can think of optimizing this structure to be more performant and memory efficient by adopting alternative implementation (eg [judy arrays](https://judy.sourceforge.net/doc/10minutes.htm) and other proposed implementations) this work might still not yield better memory efficiency and will require long and complicated implementation.
+
+
+### 2. Skiplist buckets 
+
+Since we are introducing object expiration support for hash objects, we cannot allow sorting all items by their TTL since this would imply logarithmic complexity on HashTable operations. We could, however,  attempt to achieve approximate constant complexity when modifying Objects. This is done by bounding the number of items stored in a sorted data structure.
+
+In order to manage Objects in a sorted data structure, we would first need to  extend the object metadata:
+
+![Alt text](images/HFE-images/image14.png)
+
+Note that we added the Next pointer to the object metadata (now each object overhead is **16 bytes** when expiration time is set). We will manage the objects in a data structure known as a deterministic skip-list. Unlike regular skip-list which uses random choosing for the size of the “fat-keys”, deterministic skip list is always reconstructed to support the optimal list (or tree) structure. For example deterministic skip lists are isomorphism to the Btree family of trees  . Specifically [B+ trees](https://en.wikipedia.org/wiki/B%2B_tree)can usually be implemented using a deterministic skip list. 
+
+![Alt text](images/HFE-images/image15.png)
+
+#### How do we limit the number of objects?
+
+As mentioned before, we cannot allow logarithmic complexity. this means we will have to keep the list small enough (say 512 elements). But what happens when we have more than 512 elements with expiry data?
+For that we will manage The RAX data structure mentioned before. Whenever a skip list has grown over the maximum amount of Objects we will split it to 2 lists. SkipLists are great in supporting range split and merges. We can locate the median value in a logarithmic time and since the list is already sorted.
+
+### Active expiry cycle and credits 
+
+We plan to introduce a new type of active expiry cycle (in addition to `ACTIVE_EXPIRE_CYCLE_FAST` and `ACTIVE_EXPIRE_CYCLE_SLOW` ): `ACTIVE_EXPIRE_CYCLE_ITEMS`.
+The new expiry cycle will use the same overall logic as the regular active expiry cycle with the following adjustments:  
```

Comment:
Key active expiry is built around iterating over databases, and for each DB it scans db->expires using a hit-or-miss approach over keys with expiry. Hash field expiry doesn’t follow the same model, it’s not based on random sampling, and its pacing is entirely different.

I think trying to couple field expiry with key expiry would require a significant refactor, and I don’t see a clear benefit. Expiring keys and fields are logically unrelated and should remain decoupled. I don’t see how field expiry fits naturally into the existing loops in `activeExpireCycle()`. Entry expiry should have its own database iterator IMO.

Using the same logic for both means combining two almost entirely independent mechanisms into one, which adds complexity without meaningful gain. Is it really worth it?

Reply:
Separate sounds good to me, unless Ran has some good motivation for coupling them.